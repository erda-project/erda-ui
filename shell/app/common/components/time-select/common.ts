// Copyright (c) 2021 Terminus, Inc.
//
// This program is free software: you can use, redistribute, and/or modify
// it under the terms of the GNU Affero General Public License, version 3
// or later ("AGPL"), as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

import i18n from 'i18n';
import moment, { Moment } from 'moment';

export type IRelativeTime = keyof typeof relativeTimeRange;

export type IRefreshDuration = keyof typeof autoRefreshDuration;

export interface ITimeRange {
  mode: 'quick' | 'customize';
  quick?: IRelativeTime;
  customize: {
    start?: Moment;
    end?: Moment;
  };
}

export const defaultFormat = 'YYYY-MM-DD HH:mm:ss';

export const relativeTimeRange = {
  'minutes:15': i18n.t('last {amount} {unit}', { amount: 15, unit: i18n.t('common:minutes') }),
  'minutes:30': i18n.t('last {amount} {unit}', { amount: 30, unit: i18n.t('common:minutes') }),
  'hours:1': i18n.t('last {amount} {unit}', { amount: 1, unit: i18n.t('hours') }),
  'hours:3': i18n.t('last {amount} {unit}', { amount: 3, unit: i18n.t('hours') }),
  'hours:6': i18n.t('last {amount} {unit}', { amount: 6, unit: i18n.t('hours') }),
  'hours:12': i18n.t('last {amount} {unit}', { amount: 12, unit: i18n.t('hours') }),
  'days:1': i18n.t('last {amount} {unit}', { amount: 1, unit: i18n.t('days') }),
  'days:3': i18n.t('last {amount} {unit}', { amount: 3, unit: i18n.t('days') }),
  'weeks:1': i18n.t('last {amount} {unit}', { amount: 1, unit: i18n.t('weeks') }),
  'months:1': i18n.t('last {amount} {unit}', { amount: 1, unit: i18n.t('months') }),
  'months:3': i18n.t('last {amount} {unit}', { amount: 3, unit: i18n.t('months') }),
  'months:6': i18n.t('last {amount} {unit}', { amount: 6, unit: i18n.t('months') }),
  today: i18n.t('today'),
  yesterday: i18n.t('yesterday'),
  currentWeek: i18n.t('current week'),
  lastWeek: i18n.t('last week'),
  currentMonth: i18n.t('current month'),
  lastMonth: i18n.t('last month'),
};

export const autoRefreshDuration = {
  'seconds:5': '5s',
  'seconds:10': '10s',
  'seconds:30': '30s',
  'minutes:1': '1m',
  'minutes:5': '5m',
  'minutes:30': '30m',
  'hours:1': '1h',
  'hours:2': '2h',
  'days:1': '1d',
};

/**
 * @description convert relative time range to absolute time range
 * @param unit
 * @param count
 * @returns {[Moment, Moment]}
 */
export const translateRelativeTime = (unit: string, count?: number) => {
  let start = moment();
  let end = start.clone();
  switch (unit) {
    case 'minutes':
    case 'hours':
    case 'days':
    case 'weeks':
    case 'months':
      start = start.subtract(count, unit);
      break;
    case 'today':
      start = start.startOf('day');
      break;
    case 'yesterday':
      start = start.startOf('day').subtract(1, 'days');
      end = end.endOf('day').subtract(1, 'days');
      break;
    case 'currentWeek':
      start = start.startOf('week');
      break;
    case 'lastMonth':
      start = start.startOf('month').subtract(1, 'months');
      end = end.subtract(1, 'months').endOf('month');
      break;
    case 'lastWeek':
      start = start.startOf('week').subtract(1, 'weeks');
      end = end.endOf('week').subtract(1, 'weeks');
      break;
    case 'currentMonth':
      start = start.startOf('month');
      break;
    default:
      start = start.startOf('day');
      break;
  }
  return [start, end];
};

/**
 * @description calculate the auto refresh duration
 * @param count
 * @param unit
 * @return {number}
 */
export const translateAutoRefreshDuration = (count: number, unit: string) => {
  const map = {
    seconds: 1000,
    minutes: 60 * 1000,
    hours: 60 * 60 * 1000,
    days: 24 * 60 * 60 * 1000,
  };
  return count * map[unit];
};

export const transformRange = (v: ITimeRange, format = defaultFormat) => {
  let dateStr;
  let dateArr: Moment[] = [];
  if (v.mode === 'quick' && v.quick) {
    dateStr = relativeTimeRange[v.quick];
    const [unit, count] = v.quick?.split(':') || [];
    if (unit) {
      dateArr = translateRelativeTime(unit, parseInt(count, 10));
    }
  } else if (v.mode === 'customize') {
    const { start, end } = v.customize;
    if (start && end) {
      dateStr = `${start.format(format)} - ${end.format(format)}`;
      dateArr = [start, end];
    }
  }
  return {
    date: dateArr,
    dateStr,
  };
};
