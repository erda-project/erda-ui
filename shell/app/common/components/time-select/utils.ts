// Copyright (c) 2021 Terminus, Inc.
//
// This program is free software: you can use, redistribute, and/or modify
// it under the terms of the GNU Affero General Public License, version 3
// or later ("AGPL"), as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

import i18n from 'i18n';
import moment, { Moment } from 'moment';

export type IRelativeTime = keyof typeof relativeTimeRange;

export type IRefreshDuration = keyof typeof autoRefreshDuration;

export interface ITimeRange {
  mode: 'quick' | 'customize';
  quick?: IRelativeTime;
  customize: {
    start?: Moment;
    end?: Moment;
  };
}

export const defaultFormat = 'YYYY-MM-DD HH:mm:ss';

export const relativeTimeRange = {
  'minutes:15': `${i18n.t('last')} ${i18n.t('{num} minutes', { num: 15 })}`,
  'minutes:30': `${i18n.t('last')} ${i18n.t('{num} minutes', { num: 30 })}`,
  'hours:1': `${i18n.t('last')} ${i18n.t('{num} hour', { num: 1 })}`,
  'hours:3': `${i18n.t('last')} ${i18n.t('{num} hours', { num: 3 })}`,
  'hours:6': `${i18n.t('last')} ${i18n.t('{num} hours', { num: 6 })}`,
  'hours:12': `${i18n.t('last')} ${i18n.t('{num} hours', { num: 12 })}`,
  'days:1': `${i18n.t('last')} ${i18n.t('{num} day', { num: 1 })}`,
  'days:3': `${i18n.t('last')} ${i18n.t('{num} days', { num: 3 })}`,
  'weeks:1': `${i18n.t('last')} ${i18n.t('{num} week', { num: 1 })}`,
  'months:1': `${i18n.t('last')} ${i18n.t('{num} mouth', { num: 1 })}`,
  'months:3': `${i18n.t('last')} ${i18n.t('{num} mouths', { num: 3 })}`,
  'months:6': `${i18n.t('last')} ${i18n.t('{num} mouths', { num: 6 })}`,
  today: i18n.t('today'),
  yesterday: i18n.t('Yesterday'),
  currentWeek: i18n.t('This week'),
  lastWeek: i18n.t('Last week'),
  currentMonth: i18n.t('This month'),
  lastMonth: i18n.t('Last month'),
};

export const autoRefreshDuration = {
  'seconds:5': '5s',
  'seconds:10': '10s',
  'seconds:30': '30s',
  'minutes:1': '1m',
  'minutes:5': '5m',
  'minutes:30': '30m',
  'hours:1': '1h',
  'hours:2': '2h',
  'days:1': '1d',
};

/**
 * @description convert relative time range to absolute time range
 * @param unit
 * @param count
 * @returns {[Moment, Moment]}
 */
export const translateRelativeTime = (unit: string, count?: number) => {
  let start = moment();
  let end = start.clone();
  switch (unit) {
    case 'minutes':
    case 'hours':
    case 'days':
    case 'weeks':
    case 'months':
      start = start.subtract(count, unit);
      break;
    case 'today':
      start = start.startOf('day');
      break;
    case 'yesterday':
      start = start.startOf('day').subtract(1, 'days');
      end = end.endOf('day').subtract(1, 'days');
      break;
    case 'currentWeek':
      start = start.startOf('week');
      break;
    case 'lastMonth':
      start = start.startOf('month').subtract(1, 'months');
      end = end.subtract(1, 'months').endOf('month');
      break;
    case 'lastWeek':
      start = start.startOf('week').subtract(1, 'weeks');
      end = end.endOf('week').subtract(1, 'weeks');
      break;
    case 'currentMonth':
      start = start.startOf('month');
      break;
    default:
      start = start.startOf('day');
      break;
  }
  return [start, end];
};

/**
 * @description calculate the auto refresh duration
 * @param count
 * @param unit
 * @return {number}
 */
export const translateAutoRefreshDuration = (count: number, unit: string) => {
  const map = {
    seconds: 1000,
    minutes: 60 * 1000,
    hours: 60 * 60 * 1000,
    days: 24 * 60 * 60 * 1000,
  };
  return count * map[unit];
};

export const transformRange = (v: ITimeRange, format = defaultFormat) => {
  let dateStr;
  let dateArr: Moment[] = [];
  if (v.mode === 'quick' && v.quick) {
    dateStr = relativeTimeRange[v.quick];
    const [unit, count] = v.quick?.split(':') || [];
    if (unit) {
      dateArr = translateRelativeTime(unit, parseInt(count, 10));
    }
  } else if (v.mode === 'customize') {
    const { start, end } = v.customize;
    if (start && end) {
      dateStr = `${start.format(format)} - ${end.format(format)}`;
      dateArr = [start, end];
    }
  }
  return {
    date: dateArr,
    dateStr,
  };
};
