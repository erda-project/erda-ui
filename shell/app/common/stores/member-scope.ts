// Copyright (c) 2021 Terminus, Inc.
//
// This program is free software: you can use, redistribute, and/or modify
// it under the terms of the GNU Affero General Public License, version 3
// or later ("AGPL"), as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

// if MemberScope is exported in 'common/stores/_member', jest runtime will get the error "Cannot read property 'ORGâ€™ of undefined"
// it is really confusing
export enum MemberScope {
  ORG = 'org',
  PROJECT = 'project',
  APP = 'app',
  MSP = 'msp',
  SYS = 'sys',
}
