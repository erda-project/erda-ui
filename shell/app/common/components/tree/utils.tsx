// Copyright (c) 2021 Terminus, Inc.
//
// This program is free software: you can use, redistribute, and/or modify
// it under the terms of the GNU Affero General Public License, version 3
// or later ("AGPL"), as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

import React from 'react';
import { find, set } from 'lodash';
import { TreeNode } from './tree';
import { Icon as CustomIcon } from 'common';

/**
 * 在treeData中遍历找到目标节点
 * @param nodeKey
 */
export const findTargetNode = (nodeKey: string, nodeTree: TreeNode[]): null | TreeNode => {
  if (nodeTree && nodeTree.length > 0) {
    let target = find(nodeTree, { key: nodeKey }) || null;
    if (!target) {
      let i = 0;
      while (!target && i < nodeTree.length) {
        const { children } = nodeTree[i];
        if (children) {
          target = findTargetNode(nodeKey, children);
        }
        i += 1;
      }
    }
    return target;
  }
  return null;
};

/**
 * 测试parentKey是否是childKey的直系祖先
 * 使用深度优先遍历从树顶往下找
 * 如果先找到childKey，判断parentKey是否在它的子孙上，如果在子孙里能找到返回false，否则true
 * 如果先找到parentKey，判断childKey是否在它的子孙上，如果能找到返回true，否则false
 * 如果都没找到返回false
 * 找到任意一个点，就只遍历这个点的子孙，剩下的兄弟不处理
 * @param nodeTree
 * @param childKey
 * @param parentKey
 */
export const isAncestor = (nodeTree: TreeNode[], childKey?: string, parentKey?: string| null) => {
  if (!parentKey || !childKey) {
    return false;
  }
  if (parentKey === childKey) {
    return false;
  }
  let result = false;
  let end = false;
  let capturedParent = false;
  let capturedChild = false;
  const root = nodeTree[0];
  const deepLoop = (node: TreeNode) => {
    if (end) {
      return;
    }
    if (node.key === childKey) {
      capturedChild = true;
      if (capturedParent) {
        result = true;
        end = true;
      }
    } else if (node.key === parentKey) {
      capturedParent = true;
      if (capturedChild) {
        result = false;
        end = true;
      }
    }
    const { children } = node;
    if (children && children.length > 0) {
      for (let i = 0; i < children.length; i++) {
        deepLoop(children[i]);
      }
      if (capturedParent !== capturedChild) { // 找到任意一个点，其子孙遍历完，不管有没有找到另一个都结束
        end = true;
      }
    } else if (capturedParent !== capturedChild) {
      end = true;
    }
  };
  deepLoop(root);
  return result;
};

/**
 * 递归遍历树，将options中的属性赋给所有节点，除了excludeKey
 * @param treeData
 * @param options
 * @param excludeKey
 */
export const walkTree = (treeData: TreeNode[], options: { disabled?: boolean }, excludeKey: string) => {
  treeData.forEach((node) => {
    Object.keys(options).forEach((optionKey) => {
      if (excludeKey === node.key) {
        return;
      }
      set(node, optionKey, options[optionKey]);
      if (node.children) {
        walkTree(node.children, options, excludeKey);
      }
    });
  });
};

const defaultFolderIcon = () => <CustomIcon color type="folder" style={{ height: '16px' }} />;
const defaultFileIcon = () => <CustomIcon color type="dm" style={{ height: '16px' }} />;

/**
 * 获取指定类型或者默认的图标
 * @param param
 * @param iconMap
 */
export const getIcon = ({ isLeaf, type, iconType }: { isLeaf?: boolean; type?: string; iconType?: string }, iconMap?: { [p: string]: JSX.Element }): JSX.Element => {
  if (iconMap) {
    if (iconType && iconMap[iconType]) {
      return iconMap[iconType];
    } else if (type && iconMap[type]) {
      return iconMap[type];
    }
  }
  return isLeaf ? defaultFileIcon() : defaultFolderIcon();
};

/**
 * 将后端标准文件节点类型转化成组件Node类型
 * @param node
 */
export const convertNodeType = (node: TREE.NODE): TreeNode => {
  const { inode, name, type, pinode } = node;
  return { key: inode, title: name, titleAlias: name, isLeaf: type === 'f', parentKey: pinode, originData: node };
};

/**
 * 给文件夹列表排序，文件夹在前，文件在后，按文件名称升序排列
 * @param nodes
 */
export const sortCategory = (nodes: TreeNode[]) => {
  const folders = [...nodes.filter((node) => !node.isLeaf)];
  const files = [...nodes.filter((node) => node.isLeaf)];
  const collator = new Intl.Collator(undefined, { numeric: true, sensitivity: 'base' });
  folders.sort((x, y) => collator.compare(x.titleAlias, y.titleAlias));
  files.sort((x, y) => collator.compare(x.titleAlias, y.titleAlias));
  return [...folders, ...files];
};

