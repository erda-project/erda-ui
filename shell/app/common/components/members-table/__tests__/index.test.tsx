// Copyright (c) 2021 Terminus, Inc.
//
// This program is free software: you can use, redistribute, and/or modify
// it under the terms of the GNU Affero General Public License, version 3
// or later ("AGPL"), as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.
import React from 'react';
import MembersTable from '../index';
import { act, fireEvent, render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';

import { MemberScope } from 'common/stores/member-scope';
import memberLabelStore from 'common/stores/member-label';
import userStore from 'user/stores';
import orgStore from 'app/org-home/stores/org';
import orgMemberStore from 'common/stores/org-member';
import appMemberStore from 'common/stores/application-member';
import { getDefaultPaging } from 'common/utils';
import _ from 'lodash';
import { message } from 'antd';
import ErdaTable from 'common/components/table';
import FormModal from 'common/components/form-modal';
import BatchAuthorizeMemberModal from 'common/components/batch-authorize-member-modal';
import AuthorizeMemberModal from 'common/components/authorize-member-modal';
import AddMemberModal from 'common/components/add-member-modal';

jest.mock('common/components/table');
jest.mock('common/components/form-modal');
jest.mock('common/components/batch-authorize-member-modal');
jest.mock('common/components/authorize-member-modal');
jest.mock('common/components/add-member-modal');

const memberLabels = [
  {
    label: 'Outsource',
    name: '外包人员',
  },
  {
    label: 'Partner',
    name: '合作伙伴',
  },
];

const currentOrg = {
  id: '27',
  creator: '12345t',
  desc: '',
  logo: '',
  name: 'cdp',
  displayName: 'dice-dev-pro',
  locale: '',
  isPublic: true,
  enableReleaseCrossCluster: false,
  selected: false,
  operation: '',
  status: '',
  type: 'FREE',
  publisherId: 0,
  domain: 'cdp-org.dev.terminus.io',
  version: 0,
  createdAt: '2021-06-21T09:43:02+08:00',
  updatedAt: '2021-06-21T09:43:02+08:00',
};

const loginUser = {
  id: '12345t',
  email: '',
  nick: 'erda',
  name: 'ERDA',
  phone: '',
  avatar: '',
  token: 'abc',
  isSysAdmin: false,
};

const list: IMember[] = [
  {
    id: '12345t',
    userId: '12345t',
    email: 'erda@erda.cloud',
    mobile: '123****5294',
    name: 'ERDA',
    nick: 'erda',
    avatar: '',
    status: '',
    scope: {
      type: 'org',
      id: '27',
    },
    roles: ['Lead'],
    labels: null,
    removed: false,
  },
  {
    id: '12345o',
    userId: '12345o',
    email: 'dice@erda.cloud',
    mobile: '124****5294',
    name: 'DICE',
    nick: 'dice',
    avatar: '',
    status: '',
    scope: {
      type: 'org',
      id: '27',
    },
    roles: ['Lead'],
    labels: null,
    removed: false,
  },
];

const roleMap = {
  Owner: 'Owner',
  Lead: 'Lead',
  PM: 'PM',
  PD: 'PD',
};

describe('MembersTable', () => {
  beforeAll(() => {
    jest.mock('common/stores/member-label');
    jest.mock('app/user/stores');
    jest.mock('app/org-home/stores/org');
    jest.mock('common/stores/org-member');
    jest.mock('common/stores/application-member');
    jest.mock('lodash');
    memberLabelStore.useStore = (fn) => fn({ memberLabels });
    orgStore.useStore = (fn) => fn({ currentOrg });
    userStore.useStore = (fn) => fn({ loginUser });
    orgMemberStore.useStore = (fn) => fn({ list, roleMap, paging: getDefaultPaging() });
    appMemberStore.effects.getRoleMap = jest.fn();
    _.debounce = (fn: Function) => fn;
  });
  afterAll(() => {
    jest.resetAllMocks();
  });
  it('MembersTable should work well', async () => {
    const spyError = jest.spyOn(message, 'error');
    jest.useFakeTimers();
    const getMemberLabels = jest.fn();
    const getRoleMap = jest.fn();
    const getMemberList = jest.fn();
    const cleanMembers = jest.fn();
    const updateMembers = jest.fn();
    const removeMember = jest.fn().mockResolvedValue();
    const genOrgInviteCode = jest.fn().mockResolvedValue({ verifyCode: 'verifyCode' });
    memberLabelStore.effects.getMemberLabels = getMemberLabels;
    orgMemberStore.effects.getRoleMap = getRoleMap;
    orgMemberStore.effects.updateMembers = updateMembers;
    orgMemberStore.effects.getMemberList = getMemberList;
    orgMemberStore.effects.removeMember = removeMember;
    orgMemberStore.effects.genOrgInviteCode = genOrgInviteCode;
    orgMemberStore.reducers.cleanMembers = cleanMembers;
    const paginationChange = jest.fn();
    const rowClassFn = jest.fn();
    let formModalProps: Record<string, any> = {};
    let addMemberProps: Record<string, any> = {};
    let tableActions: { render: (data: IMember) => Array<any> } = {
      render: () => [],
    };
    (ErdaTable as unknown as jest.MockInstance<any, any>).mockImplementation(
      ({ slot, pagination, rowClassName, ...rest }) => {
        paginationChange.mockImplementation(() => {
          pagination.onChange();
        });
        rowClassFn.mockImplementation(rowClassName);
        tableActions = rest.actions || {};
        return <div className="mock-erda-table">{slot}</div>;
      },
    );
    (FormModal.render as unknown as jest.MockInstance<any, any>).mockImplementation((props) => {
      formModalProps = props;
      return (
        <div className="mock-form-modal" id={props.title}>
          {props.visible}
        </div>
      );
    });
    (BatchAuthorizeMemberModal as unknown as jest.MockInstance<any, any>).mockImplementation((props) => {
      return (
        <div className="mock-batch-authorize-member-modal" id={props.title}>
          {props.visible}
        </div>
      );
    });
    (AuthorizeMemberModal as unknown as jest.MockInstance<any, any>).mockImplementation((props) => {
      return (
        <div className="mock-authorize-member-modal" id={props.title}>
          {props.visible}
        </div>
      );
    });
    (AddMemberModal as unknown as jest.MockInstance<any, any>).mockImplementation((props) => {
      addMemberProps = props;
      return (
        <div className="mock-add-member-modal" id={props.title}>
          {props.visible}
        </div>
      );
    });
    const wrapper = render(
      <MembersTable showAuthorize scopeKey={MemberScope.ORG} overwriteAuth={{ showAuthorize: true }} />,
    );
    expect(getMemberLabels).toHaveBeenLastCalledWith();
    expect(getRoleMap).toHaveBeenLastCalledWith({ scopeId: currentOrg.id, scopeType: 'org' });
    act(() => {
      paginationChange.getMockImplementation()?.();
      expect(rowClassFn.getMockImplementation()?.({ removed: true })).toBe('not-allowed');
      expect(rowClassFn.getMockImplementation()?.({ removed: false })).toBe('');
    });
    const actions = tableActions.render?.(list[0]);
    expect(wrapper.baseElement).toMatchSnapshot();
    const [edit, authorize, remove] = actions;
    act(() => {
      edit.onClick?.();
    });
    expect(formModalProps.visible).toBeTruthy();
    act(() => {
      formModalProps.onCancel();
    });
    expect(formModalProps.visible).toBeFalsy();
    act(() => {
      remove.onClick?.();
    });
    await waitFor(() => expect(wrapper.baseElement.querySelector('.ant-modal-confirm-confirm')).toBeInTheDocument());
    fireEvent.click(screen.getByText('OK'));
    expect(removeMember).toHaveBeenCalledTimes(1);

    fireEvent.click(screen.getByText('add member'));
    expect(addMemberProps.visible).toBeTruthy();
    act(() => {
      addMemberProps.toggleModal();
    });
    expect(addMemberProps.visible).toBeFalsy();
    // expect(tableActions.length).toBe(1)
    // triggerDropdownSelect(wrapper, 'edit');
    // expect(wrapper.find({ title: 'batch set the role of member' })).toExist();
    // act(() => {
    //   wrapper
    //     .find('Table')
    //     .at(0)
    //     .prop('rowSelection')
    //     .onChange(list.map((t) => t.userId));
    // });
    // act(() => {
    //   wrapper.find({ title: 'batch set the role of member' }).at(0).prop('onOk')();
    // });
    // wrapper.update();
    // expect(updateMembers).toHaveBeenLastCalledWith(
    //   {
    //     scope: {
    //       id: '27',
    //       type: 'org',
    //     },
    //     userIds: ['12345t', '12345o'],
    //   },
    //   { isSelf: true, queryParams: {} },
    // );
    // expect(wrapper.find({ title: 'batch set the role of member' })).not.toExist();
    // triggerDropdownSelect(wrapper, 'authorize');
    // expect(wrapper.find('BatchAuthorizeMemberModal').prop('visible')).toBeTruthy();
    // act(() => {
    //   wrapper.find('BatchAuthorizeMemberModal').prop('onOk')({ applications: [1, 2], roles: ['PM', 'PD'] });
    // });
    // expect(updateMembers).toHaveBeenLastCalledWith({
    //   roles: ['PM', 'PD'],
    //   scope: { id: undefined, type: 'project' },
    //   targetScopeIds: [1, 2],
    //   targetScopeType: 'app',
    //   userIds: ['12345t', '12345o'],
    // });
    // triggerDropdownSelect(wrapper, 'remove');
    // jest.runAllTimers();
    // $$<HTMLButtonElement>('.ant-btn-primary')[2].click();
    // expect(removeMember).toHaveBeenCalledTimes(1);
    // act(() => {
    //   wrapper.find('FilterGroup').prop('onChange')({ query: 'erda-cloud', queryRole: 'admin', label: ['PM', 'PD'] });
    // });
    // triggerTableAction(wrapper, 0, 'edit');
    // act(() => {
    //   wrapper.find({ title: 'set the role of member erda' }).at(0).prop('onOk')();
    // });
    // expect(updateMembers).toHaveBeenLastCalledWith(
    //   {
    //     scope: {
    //       id: '27',
    //       type: 'org',
    //     },
    //     userIds: ['12345t'],
    //   },
    //   {
    //     isSelf: true,
    //     queryParams: {
    //       label: ['PM', 'PD'],
    //       pageNo: 1,
    //       q: 'erda-cloud',
    //       roles: ['admin'],
    //     },
    //   },
    // );
    // triggerTableAction(wrapper, 0, 'exit');
    // jest.runAllTimers();
    // $$<HTMLButtonElement>('.ant-btn-primary')[2].click();
    // expect(removeMember).toHaveBeenCalledTimes(2);
    // act(() => {
    //   wrapper.find('.members-list').find('Button').at(0).simulate('click');
    // });
    // wrapper.update();
    // expect(wrapper.find('AddMemberModal').prop('visible')).toBeTruthy();
    // act(() => {
    //   wrapper.find('AddMemberModal').prop('toggleModal')();
    // });
    // wrapper.update();
    // expect(wrapper.find('AddMemberModal').prop('visible')).toBeFalsy();
    // await act(async () => {
    //   await wrapper.find('.members-list').find('Button').at(1).simulate('click');
    // });
    // wrapper.update();
    // expect(genOrgInviteCode).toHaveBeenCalled();
    // expect(wrapper.find('UrlInviteModal').prop('visible')).toBeTruthy();
    // wrapper.find('UrlInviteModal').prop('onCancel')();
    // wrapper.update();
    // expect(wrapper.find('UrlInviteModal').prop('visible')).toBeFalsy();

    wrapper.unmount();
    expect(cleanMembers).toHaveBeenCalledTimes(1);
    spyError.mockReset();
  });
});
