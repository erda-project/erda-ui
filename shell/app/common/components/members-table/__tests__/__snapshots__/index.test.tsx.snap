// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MembersTable MembersTable should work well 1`] = `
<body
  lang="en"
>
  <script />
  <div>
    <div
      class="member-table-manage"
    >
      <div
        class="top-button-group"
      >
        <div>
          <button
            class="ant-btn ant-btn-primary"
            type="button"
          >
            <span>
              add member
            </span>
          </button>
          <button
            class="ant-btn"
            type="button"
          >
            <span>
              invite
            </span>
          </button>
        </div>
      </div>
      <div
        class="ant-spin-nested-loading"
      >
        <div
          class="ant-spin-container"
        >
          <div
            class="member-table-manage-list"
          >
            <div
              class="mock-add-member-modal"
            />
            <div
              class="mock-form-modal"
              id="set the role of member null"
            />
            <div
              class="mock-batch-authorize-member-modal"
            />
            <div
              class="mock-authorize-member-modal"
            />
            <div
              class="members-list"
            >
              <div
                class="mock-erda-table"
              >
                <div
                  class="filter-group-bar flex justify-between items-center"
                >
                  <div
                    class="filter-group-left flex-1"
                  >
                    <input
                      class="ant-input"
                      name="query"
                      placeholder="search by nickname, username, email or mobile number"
                      style="width: 260px;"
                      type="text"
                      value=""
                    />
                    <div
                      class="ant-select ant-select-single ant-select-allow-clear ant-select-show-arrow"
                      name="queryRole"
                    >
                      <div
                        class="ant-select-selector"
                      >
                        <span
                          class="ant-select-selection-search"
                        >
                          <input
                            aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
                            aria-autocomplete="list"
                            aria-controls="rc_select_TEST_OR_SSR_list"
                            aria-haspopup="listbox"
                            aria-owns="rc_select_TEST_OR_SSR_list"
                            autocomplete="off"
                            class="ant-select-selection-search-input"
                            id="rc_select_TEST_OR_SSR"
                            readonly=""
                            role="combobox"
                            style="opacity: 0;"
                            type="search"
                            unselectable="on"
                            value=""
                          />
                        </span>
                        <span
                          class="ant-select-selection-placeholder"
                        >
                          select role
                        </span>
                      </div>
                      <span
                        aria-hidden="true"
                        class="ant-select-arrow"
                        style="user-select: none;"
                        unselectable="on"
                      >
                        <span
                          aria-label="down"
                          class="anticon anticon-down ant-select-suffix"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="down"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                            />
                          </svg>
                        </span>
                      </span>
                    </div>
                    <div
                      class="ant-select ant-select-multiple ant-select-allow-clear ant-select-show-search"
                      name="label"
                    >
                      <div
                        class="ant-select-selector"
                      >
                        <div
                          class="ant-select-selection-overflow"
                        >
                          <div
                            class="ant-select-selection-overflow-item ant-select-selection-overflow-item-suffix"
                            style="opacity: 1;"
                          >
                            <div
                              class="ant-select-selection-search"
                              style="width: 0px;"
                            >
                              <input
                                aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
                                aria-autocomplete="list"
                                aria-controls="rc_select_TEST_OR_SSR_list"
                                aria-haspopup="listbox"
                                aria-owns="rc_select_TEST_OR_SSR_list"
                                autocomplete="off"
                                class="ant-select-selection-search-input"
                                id="rc_select_TEST_OR_SSR"
                                readonly=""
                                role="combobox"
                                style="opacity: 0;"
                                type="search"
                                unselectable="on"
                                value=""
                              />
                              <span
                                aria-hidden="true"
                                class="ant-select-selection-search-mirror"
                              >
                                
                                Â 
                              </span>
                            </div>
                          </div>
                        </div>
                        <span
                          class="ant-select-selection-placeholder"
                        >
                          select member label
                        </span>
                      </div>
                    </div>
                  </div>
                  <div
                    class="filter-group-right ml-6"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
